common/common-rest/src/main/java/io/servicecomb/common/rest/AbstractRestInvocation.java

```diff
@@ -0,0 +1,160 @@
+ */
+
+package io.servicecomb.common.rest;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.Response.Status;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.buffer.Unpooled;
+import io.servicecomb.common.rest.codec.produce.ProduceProcessor;
+import io.servicecomb.common.rest.codec.produce.ProduceProcessorManager;
+import io.servicecomb.common.rest.definition.RestOperationMeta;
+import io.servicecomb.common.rest.filter.HttpServerFilter;
+import io.servicecomb.core.Const;
+import io.servicecomb.core.Invocation;
+import io.servicecomb.foundation.common.utils.JsonUtils;
+import io.servicecomb.foundation.vertx.http.HttpServletRequestEx;
+import io.servicecomb.foundation.vertx.http.HttpServletResponseEx;
+import io.servicecomb.foundation.vertx.stream.BufferOutputStream;
+import io.servicecomb.swagger.invocation.Response;
+import io.servicecomb.swagger.invocation.exception.InvocationException;
+
+public abstract class AbstractRestInvocation {
+  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractRestInvocation.class);
+
+  protected RestOperationMeta restOperationMeta;
+
+  protected Invocation invocation;
+
+  protected HttpServletRequestEx requestEx;
+
+  protected HttpServletResponseEx responseEx;
+
+  protected ProduceProcessor produceProcessor;
+
+  protected List<HttpServerFilter> httpServerFilters = Collections.emptyList();
+
+  public void setHttpServerFilters(List<HttpServerFilter> httpServerFilters) {
+    this.httpServerFilters = httpServerFilters;
+  }
+
+  protected boolean initProduceProcessor() {
+    produceProcessor = restOperationMeta.ensureFindProduceProcessor(requestEx);
+    if (produceProcessor == null) {
+      String msg = String.format("Accept %s is not supported", requestEx.getHeader(HttpHeaders.ACCEPT));
+      InvocationException exception = new InvocationException(Status.NOT_ACCEPTABLE, msg);
+      sendFailResponse(exception);
+      return false;
+    }
+
+    return true;
+  }
+
+  protected void setContext() throws Exception {
+    String strCseContext = requestEx.getHeader(Const.CSE_CONTEXT);
+    if (StringUtils.isEmpty(strCseContext)) {
+      return;
+    }
+
+    @SuppressWarnings("unchecked")
+    Map<String, String> cseContext =
+        JsonUtils.readValue(strCseContext.getBytes(StandardCharsets.UTF_8), Map.class);
+    invocation.setContext(cseContext);
+  }
+
+  protected void invoke() {
+    try {
+      this.initProduceProcessor();
+      this.setContext();
+      invocation.getHandlerContext().put(RestConst.REST_REQUEST, requestEx);
+
+      for (HttpServerFilter filter : httpServerFilters) {
+        Response response = filter.afterReceiveRequest(invocation, requestEx);
+        if (response != null) {
+          sendResponseQuietly(response);
```

Comment:
it looks like the filter could hijack the response message which calling the doInvoke method.
I don't think it's good idea, there are too much assumption.

Reply:
modify response is one purpose of filter definition：
  producer generator signature and write it to response header.

Reply:
Please add some comments in the HttpServerFilter.afterReceiveRequst,  to make sure user know about this kind of feature.

Reply:
already have comments in HttpServerFilter.afterReceiveRequst and HttpClientFilter.afterReceiveResponse in current PR code.

![image](https://user-images.githubusercontent.com/16874843/30890712-18b884e4-a361-11e7-83ff-e7d5a95f141c.png)

![image](https://user-images.githubusercontent.com/16874843/30890721-21222054-a361-11e7-8c5f-5cfccac5d931.png)

Reply:
OK.